test/functional/wallet_upgrade.py

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env python3
+# Copyright (c) 2014-2017 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test the walletupgrade functionality.
+
+- 1) start one pivxd node from an pre-HD wallet wallet.dat file localized in util/data
+- 2) dumpwallet and get all the keys.
+- 3) Upgrade Wallet: stopnode and initialize it with the -upgradewallet flag.
+- 4) Verify that all of the pre-HD keys are still in the upgraded wallet.
+- 5) Dry out the pre-HD keypool.
+- 6) Generate new addresses and verify HD path correctness.
+"""
+import os
+import shutil
+from test_framework.test_framework import PivxTestFramework
+from test_framework.util import (
+    assert_equal,
+    assert_raises_rpc_error,
+    wait_until,
+    initialize_datadir,
+)
+
+def read_dump(file_name, addrs):
+    """
+    Read the given dump, count the addrs that match, count change and reserve.
+    Also check that the old hd_master is inactive
+    """
+    with open(file_name, encoding='utf8') as inputfile:
+        found_addr = 0
+        found_addr_rsv = 0
+        for line in inputfile:
+            # only read non comment lines
+            if line[0] != "#" and len(line) > 10:
+                # split out some data
+                key_date_label, comment = line.split("#")
+                key_date_label = key_date_label.split(" ")
+
+                date = key_date_label[1]
+                keytype = key_date_label[2]
+
+                imported_key = date == '1970-01-01T00:00:01Z'
+                if imported_key:
+                    # Imported keys have multiple addresses, no label (keypath) and timestamp
+                    # Skip them
+                    continue
+
+                addr_keypath = comment.split(" addr=")[1]
+                addr = addr_keypath.split(" ")[0]
+
+                if keytype == "reserve=1":
+                    found_addr_rsv += 1
+                else:
+                    found_addr += 1
+
+                # Ret addresses
+                addrs.append(addr)
+
+        return found_addr, found_addr_rsv
+
+class WalletUpgradeTest (PivxTestFramework):
+
+    def _initialize_chain_clean(self):
+        initialize_datadir(self.options.tmpdir, 0)
+        destinationDirPath = os.path.join(self.options.tmpdir, "node0", "regtest")
+        os.makedirs(destinationDirPath, exist_ok=False)
```

Comment:
`_initialize_chain_clean` is not meant to be overridden.
You can override `setup_chain()` from `PivxTestFramework`, and call `_initialize_chain_clean()` from there (that will take care of creating the datadir, no need to do it here). Then you can copy the pre-hd wallet file.
`setup_chain` is called by the main test function before `setup_network` (which is called before `run_test()`)

Reply:
k, will refactor it.

Reply:
done